Lightning CLI: missing 1 required keyword-only argument: applied_instantiation_links

### Description

The lightning extras [require](https://github.com/Lightning-AI/pytorch-lightning/blob/master/requirements/pytorch/extra.txt#L8) `jsonargparse >= 4.39` but this is not mirrored in torchgeo, resulting in errors:
```
_InstantiatorFn.__call__() missing 1 required keyword-only argument:
'applied_instantiation_links'
```
Fix: require jsonargparse >= 4.39

### Steps to reproduce

Encountered in https://github.com/torchgeo/torchgeo/pull/3625

### Version

main

```
uv pip install "lightning==2.5.2" "jsonargparse[signatures]==4.35"
.venv/bin/pytest 'tests/trainers/test_segmentation.py::TestSemanticSegmentationTask::test_trainer[True-agrifieldnet]' -v

E       ValueError: Does not validate against any of the Union subtypes
E       Subtypes: (<class 'torchgeo.trainers.base.BaseTask'>, <class 'NoneType'>)
E       Errors:
E         - _InstantiatorFn.__call__() missing 1 required keyword-only argument: 'applied_instantiation_links'
E         - Expected a <class 'NoneType'>
E       Given value type: <class 'jsonargparse._namespace.Namespace'>
E       Given value: Namespace(class_path='torchgeo.trainers.SemanticSegmentationTask', init_args=Namespace(model='unet', backbone='resnet18', weights=None, in_channels=12, task='multiclass', num_classes=14, num_labels=None, labels=None, num_filters=1, pos_weight=None, loss='ce', class_weights=None, ignore_index=0, lr=0.001, patience=10, freeze_backbone=False, freeze_decoder=False))

.venv/lib/python3.12/site-packages/jsonargparse/_typehints.py:710: ValueError
```

--- a/src/lightning/pytorch/cli.py
+++ b/src/lightning/pytorch/cli.py
@@ -36,7 +36,7 @@ from lightning.pytorch.utilities.exceptions import MisconfigurationException
 from lightning.pytorch.utilities.model_helpers import is_overridden
 from lightning.pytorch.utilities.rank_zero import rank_zero_warn
 
-_JSONARGPARSE_SIGNATURES_AVAILABLE = RequirementCache("jsonargparse[signatures]>=4.27.7")
+_JSONARGPARSE_SIGNATURES_AVAILABLE = RequirementCache("jsonargparse[jsonnet,signatures]>=4.39")
 
 if _JSONARGPARSE_SIGNATURES_AVAILABLE:
     import docstring_parser

--- a/tests/tests_pytorch/test_cli.py
+++ b/tests/tests_pytorch/test_cli.py
@@ -259,7 +259,7 @@ def test_lightning_cli_args(cleandir):
     assert loaded_config["trainer"] == cli_config["trainer"]
 
 
-@pytest.mark.skipif(compare_version("jsonargparse", operator.lt, "4.21.3"), reason="vulnerability with failing imports")
+@pytest.mark.skipif(compare_version("jsonargparse", operator.lt, "4.39"), reason="incompatibilities with older jsonargparse")
 def test_lightning_env_parse(cleandir):
     out = StringIO()
     with mock.patch("sys.argv", ["", "fit", "--help"]), redirect_stdout(out), pytest.raises(SystemExit):
@@ -426,12 +426,7 @@ def any_model_any_data_cli():
     LightningCLI(LightningModule, LightningDataModule, subclass_mode_model=True, subclass_mode_data=True)
 
 
-@pytest.mark.skipif(compare_version("jsonargparse", operator.lt, "4.21.3"), reason="vulnerability with failing imports")
-@pytest.mark.skipif(
-    (sys.version_info.major, sys.version_info.minor) == (3, 9)
-    and compare_version("jsonargparse", operator.lt, "4.24.0"),
-    reason="--trainer.precision is not parsed",
-)
+@pytest.mark.skipif(compare_version("jsonargparse", operator.lt, "4.39"), reason="incompatibilities with older jsonargparse")
 def test_lightning_cli_help():
     cli_args = ["any.py", "fit", "--help"]
     out = StringIO()
@@ -1211,7 +1206,7 @@ def test_lightning_cli_subcommands():
             assert e in parameters
 
 
-@pytest.mark.skipif(compare_version("jsonargparse", operator.lt, "4.21.3"), reason="vulnerability with failing imports")
+@pytest.mark.skipif(compare_version("jsonargparse", operator.lt, "4.39"), reason="incompatibilities with older jsonargparse")
 def test_lightning_cli_custom_subcommand():
     class TestTrainer(Trainer):
         def foo(self, model: LightningModule, x: int, y: float = 1.0):

--- a/tests/tests_pytorch/test_cli.py
+++ b/tests/tests_pytorch/test_cli.py
@@ -16,7 +16,6 @@ import inspect
 import json
 import operator
 import os
-import sys
 from contextlib import ExitStack, contextmanager, redirect_stderr, redirect_stdout
 from io import StringIO
 from pathlib import Path
@@ -259,7 +258,9 @@ def test_lightning_cli_args(cleandir):
     assert loaded_config["trainer"] == cli_config["trainer"]
 
 
-@pytest.mark.skipif(compare_version("jsonargparse", operator.lt, "4.39"), reason="incompatibilities with older jsonargparse")
+@pytest.mark.skipif(
+    compare_version("jsonargparse", operator.lt, "4.39"), reason="incompatibilities with older jsonargparse"
+)
 def test_lightning_env_parse(cleandir):
     out = StringIO()
     with mock.patch("sys.argv", ["", "fit", "--help"]), redirect_stdout(out), pytest.raises(SystemExit):
@@ -426,7 +427,9 @@ def any_model_any_data_cli():
     LightningCLI(LightningModule, LightningDataModule, subclass_mode_model=True, subclass_mode_data=True)
 
 
-@pytest.mark.skipif(compare_version("jsonargparse", operator.lt, "4.39"), reason="incompatibilities with older jsonargparse")
+@pytest.mark.skipif(
+    compare_version("jsonargparse", operator.lt, "4.39"), reason="incompatibilities with older jsonargparse"
+)
 def test_lightning_cli_help():
     cli_args = ["any.py", "fit", "--help"]
     out = StringIO()
@@ -1206,7 +1209,9 @@ def test_lightning_cli_subcommands():
             assert e in parameters
 
 
-@pytest.mark.skipif(compare_version("jsonargparse", operator.lt, "4.39"), reason="incompatibilities with older jsonargparse")
+@pytest.mark.skipif(
+    compare_version("jsonargparse", operator.lt, "4.39"), reason="incompatibilities with older jsonargparse"
+)
 def test_lightning_cli_custom_subcommand():
     class TestTrainer(Trainer):
         def foo(self, model: LightningModule, x: int, y: float = 1.0):

--- a/tests/tests_pytorch/utilities/test_imports.py
+++ b/tests/tests_pytorch/utilities/test_imports.py
@@ -87,7 +87,7 @@ def clean_import():
         ),
         pytest.param(
             "lightning_utilities.core.imports.RequirementCache.__bool__",
-            _shortcut_patch(RequirementCache.__bool__, ("jsonargparse[signatures]>=4.12.0",), ("requirement",)),
+            _shortcut_patch(RequirementCache.__bool__, ("jsonargparse[jsonnet,signatures]>=4.39",), ("requirement",)),
             "lightning.pytorch.cli",
             id="cli",
         ),

